cli/src/InterpreterApp.scala

```diff
@@ -0,0 +1,154 @@
+      .minOccurs(0)
+      .action((x, c) => c.copy(args = c.args :+ x))
+      .text("Input arguments")
+
+    opt[String]('m', "main")
+      .optional()
+      .action((f, c) => c.copy(main = f))
+      .text("Executes provided function name as main. Prints available functions if no name is provided")
+
+    opt[Boolean]('p', "parse")
+      .optional()
+      .action((f, c) => c.copy(parse = f))
+      .text("Parse the input file as text")
+
+    opt[String]('f', "filter")
+      .optional()
+      .action((f, c) => c.copy(traceFilter = f))
+      .text("Filter the traces. The parameter is a regular expression applied to the opcode, e.g.: 'mread|mwrite'")
+
+    opt[String]('l', "logfile-name")
+      .optional()
+      .action((f, c) => c.copy(tracePattern = f))
+      .text("SWAM uses a logger based on [[https://docs.oracle.com/en/java/javase/13/docs/api/java.logging/java/util/logging/package-summary.html java.util.logging]]. Provide any valid valid file pattern")
+
+    opt[Boolean]('d', "debug-compiler")
+      .optional()
+      .action((f, c) => c.copy(debugCompiler = f))
+      .text("Activates the debug option for thee text parser")
+
+    opt[Boolean]('t', "trace")
+      .optional()
+      .action((f, c) => c.copy(trace = f))
+      .text("Traces WASM execution channels; stack, memory and opcodes")
+
+  }
+
+  def buffer = {
+    val buffer = ByteBuffer.allocate(2 * pageSize)
+    buffer.limit(pageSize)
+    buffer
+  }
+
+  def imports() =
+    Imports[IO](
+      TCMap[String, AsIsIO](
+        "env" -> TCMap[String, AsIIO]("memory" -> buffer)
+      ))
+
+  def createInstance(blocker: Blocker, config: Config): IO[Instance[IO]] = {
+    for {
+      compiler <- Compiler[IO]
+      engine <- if (config.trace)
+        Engine[IO](
+          tracer = Option(JULTracer(config.traceOutDir, config.tracePattern, NoTimestampFormatter, config.traceFilter)))
+      else
+        Engine[IO]()
+
+      module <- if (config.parse)
+        engine.compile(compiler.stream(config.wasm.toPath, config.debugCompiler, blocker))
+      else
+        engine.compile(config.wasm.toPath, blocker, 4096)
+
+      instance <- engine.instantiate(module, imports())
+    } yield instance
+  }
+
+  def run(args: List[String]): IO[ExitCode] = parser.parse(args, Config()) match {
+    case Some(config) =>
+      Blocker[IO].use { blocker =>
+        for {
+          instance <- createInstance(blocker, config)
+          _ <- if (config.main.isEmpty) {
+            IO(println("Listing available functions..."))
+            instance.exports.list
+              .collect[String, FuncType] {
+                case (name, f: FuncType) =>
+                  (name, f)
+              }
+              .foreach {
+                case (name, tpe: FuncType) => {
+                  println(s"\t ${Console.GREEN} $name ${Console.RESET}${tpe.params.mkString("( ", ",", " )")} -> ${tpe.t
+                    .mkString(",")}")
+                }
+              }
+            IO()
```

Comment:
This is not what I meant, sorry, I didn't express myself clearly enough. The operation you perform above is side-effecting (see the println), just wrap it in an `IO` like this:

```scala
IO {
  // the then code block
}
```